File: app/src/main/java/com/wuyiming/singlecolumnadfeed_android/viewmodel/FeedViewModel.java
package com.wuyiming.singlecolumnadfeed_android.viewmodel;

import android.app.Application;
import android.os.Handler;
import android.os.Looper;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.wuyiming.singlecolumnadfeed_android.ai.AiInsightGenerator;
import com.wuyiming.singlecolumnadfeed_android.ai.AiSearchService;
import com.wuyiming.singlecolumnadfeed_android.data.model.FeedCategory;
import com.wuyiming.singlecolumnadfeed_android.data.model.FeedItem;
import com.wuyiming.singlecolumnadfeed_android.data.repository.DefaultFeedRepository;
import com.wuyiming.singlecolumnadfeed_android.data.repository.FeedRepository;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.atomic.AtomicInteger;

public class FeedViewModel extends AndroidViewModel {
    private static final int PAGE_SIZE = 20;
    private static final long MIN_REFRESH_ANIMATION_MS = 600L;

    private final FeedRepository repository;
    // Three dedicated executors: refresh has its own, never blocked by AI
    private final ExecutorService dataExecutor = Executors.newSingleThreadExecutor();
    private final ExecutorService searchExecutor = Executors.newSingleThreadExecutor();
    private final ExecutorService refreshExecutor = Executors.newSingleThreadExecutor();
    private final Handler mainHandler = new Handler(Looper.getMainLooper());
    private final MutableLiveData<FeedUiState> uiState = new MutableLiveData<>();

    private int currentPage = 1;
    private FeedCategory currentCategory = FeedCategory.RECOMMEND;
    private List<FeedItem> allLoadedItems = new ArrayList<>();

    private volatile AiInsightGenerator aiGenerator;
    private volatile AiSearchService aiSearchService;
    private final AtomicInteger searchGeneration = new AtomicInteger(0);
    private final MutableLiveData<String> likeSyncEvent = new MutableLiveData<>();

    private static volatile FeedViewModel sharedInstance;

    public static FeedViewModel getSharedInstance(Application app) {
        if (sharedInstance == null) {
            synchronized (FeedViewModel.class) {
                if (sharedInstance == null) {
                    sharedInstance = new FeedViewModel(app);
                    sharedInstance.init();
                }
            }
        }
        return sharedInstance;
    }

    public static FeedViewModel getInstance(Application app) {
        return getSharedInstance(app);
    }

    public FeedViewModel(@NonNull Application application) {
        super(application);
        repository = DefaultFeedRepository.getInstance(application);
    }

    public void configureAi(String apiKey, String apiUrl, String model) {
        this.aiGenerator = new AiInsightGenerator(apiKey, apiUrl, model);
        this.aiSearchService = new AiSearchService(apiKey, apiUrl, model);
        if (!allLoadedItems.isEmpty()) {
            generateInsightsFor(allLoadedItems);
        }
    }

    private void init() {
        uiState.setValue(new FeedUiState.Builder()
                .loading(true)
                .items(new ArrayList<>())
                .currentCategory(currentCategory)
                .build());
        loadFirstPage();
    }

    private void loadFirstPage() {
        dataExecutor.execute(() -> {
            repository.reloadData();
            allLoadedItems.clear();
            currentPage = 1;
            List<FeedItem> items = repository.loadMore(currentPage, PAGE_SIZE, currentCategory);
            allLoadedItems.addAll(items);
            postFeedState(items.size() == PAGE_SIZE);
            generateInsightsFor(items);
        });
    }

    /**
     * Pull-to-refresh — runs on its own executor so it is never blocked
     * by ongoing AI generation in dataExecutor.
     */
    public void refresh() {
        final long startTime = System.currentTimeMillis();

        uiState.setValue(new FeedUiState.Builder()
                .refreshing(true)
                .items(new ArrayList<>(allLoadedItems))
                .currentCategory(currentCategory)
                .isSearchMode(false)
                .build());

        refreshExecutor.execute(() -> {
            repository.reloadData();
            allLoadedItems.clear();
            currentPage = 1;
            List<FeedItem> items = repository.loadMore(currentPage, PAGE_SIZE, currentCategory);
            allLoadedItems.addAll(items);

            long elapsed = System.currentTimeMillis() - startTime;
            long delay = Math.max(0, MIN_REFRESH_ANIMATION_MS - elapsed);

            mainHandler.postDelayed(() -> {
                uiState.setValue(new FeedUiState.Builder()
                        .loading(false)
                        .refreshing(false)
                        .items(new ArrayList<>(allLoadedItems))
                        .currentCategory(currentCategory)
                        .hasMore(items.size() == PAGE_SIZE)
                        .isSearchMode(false)
                        .build());
            }, delay);

            // Start AI generation on the data executor (may queue behind existing work)
            generateInsightsFor(items);
        });
    }

    public void loadMore() {
        dataExecutor.execute(() -> {
            uiState.postValue(new FeedUiState.Builder()
                    .loadingMore(true)
                    .items(new ArrayList<>(allLoadedItems))
                    .currentCategory(currentCategory)
                    .isSearchMode(false)
                    .build());

            currentPage++;
            List<FeedItem> items = repository.loadMore(currentPage, PAGE_SIZE, currentCategory);
            allLoadedItems.addAll(items);
            postFeedState(items.size() == PAGE_SIZE);
            generateInsightsFor(items);
        });
    }

    public void switchCategory(FeedCategory category) {
        if (currentCategory == category) return;
        currentCategory = category;
        searchGeneration.incrementAndGet();
        loadFirstPage();
    }

    public void search(String query) {
        if (query == null || query.trim().isEmpty()) {
            clearSearch();
            return;
        }

        final String trimmed = query.trim();
        final int gen = searchGeneration.incrementAndGet();

        uiState.setValue(new FeedUiState.Builder()
                .loading(false)
                .items(new ArrayList<>(allLoadedItems))
                .currentCategory(currentCategory)
                .isSearching(true)
                .build());

        searchExecutor.execute(() -> {
            if (searchGeneration.get() != gen) return;

            List<FeedItem> snapshot = new ArrayList<>(allLoadedItems);

            List<String> keywords;
            AiSearchService searchService = aiSearchService;
            if (searchService != null) {
                keywords = searchService.extractKeywords(trimmed);
            } else {
                keywords = splitLocal(trimmed);
            }

            if (searchGeneration.get() != gen) return;

            List<FeedItem> results = new ArrayList<>();
            for (FeedItem item : snapshot) {
                if (matchesKeywords(item, keywords)) {
                    results.add(item);
                }
            }

            if (searchGeneration.get() != gen) return;

            uiState.postValue(new FeedUiState.Builder()
                    .loading(false)
                    .isSearching(false)
                    .items(results)
                    .currentCategory(currentCategory)
                    .hasMore(false)
                    .isSearchMode(true)
                    .build());
        });
    }

    public void clearSearch() {
        searchGeneration.incrementAndGet();
        uiState.setValue(new FeedUiState.Builder()
                .loading(false)
                .isSearching(false)
                .items(new ArrayList<>(allLoadedItems))
                .currentCategory(currentCategory)
                .hasMore(true)
                .isSearchMode(false)
                .build());
    }

    private List<String> splitLocal(String query) {
        List<String> keywords = new ArrayList<>();
        for (String part : query.split("[\\s,.!?;:，。！？；：]+")) {
            if (part.length() >= 1) keywords.add(part);
        }
        if (keywords.isEmpty()) keywords.add(query);
        return keywords;
    }

    private boolean matchesKeywords(FeedItem item, List<String> keywords) {
        String title = item.getTitle() != null ? item.getTitle().toLowerCase() : "";
        String desc = item.getDescription() != null ? item.getDescription().toLowerCase() : "";
        StringBuilder tagStr = new StringBuilder();
        if (item.getInsight() != null && item.getInsight().getTags() != null) {
            for (String tag : item.getInsight().getTags()) {
                tagStr.append(tag.toLowerCase());
            }
        }
        String combined = title + " " + desc + " " + tagStr;

        for (String kw : keywords) {
            if (combined.contains(kw.toLowerCase())) {
                return true;
            }
        }
        return false;
    }

    private void postFeedState(boolean hasMore) {
        uiState.postValue(new FeedUiState.Builder()
                .loading(false)
                .loadingMore(false)
                .items(new ArrayList<>(allLoadedItems))
                .currentCategory(currentCategory)
                .hasMore(hasMore)
                .isSearchMode(false)
                .build());
    }

    public void toggleLike(String feedId) {
        dataExecutor.execute(() -> {
            repository.toggleLike(feedId);
            uiState.postValue(new FeedUiState.Builder()
                    .loading(false)
                    .items(new ArrayList<>(allLoadedItems))
                    .currentCategory(currentCategory)
                    .hasMore(true)
                    .build());
            likeSyncEvent.postValue(feedId);
        });
    }

    public void toggleCollect(String feedId) {
        dataExecutor.execute(() -> {
            repository.toggleCollect(feedId);
            uiState.postValue(new FeedUiState.Builder()
                    .loading(false)
                    .items(new ArrayList<>(allLoadedItems))
                    .currentCategory(currentCategory)
                    .hasMore(true)
                    .build());
        });
    }

    public FeedItem getItemById(String feedId) {
        for (FeedItem item : allLoadedItems) {
            if (item.getId().equals(feedId)) return item;
        }
        return repository.getItemById(feedId);
    }

    private void generateInsightsFor(List<FeedItem> items) {
        AiInsightGenerator gen = aiGenerator;
        if (gen == null || items == null || items.isEmpty()) return;

        dataExecutor.execute(() -> {
            for (FeedItem item : items) {
                if (item.getInsight() != null && item.getInsight().getSummary() != null) continue;
                try {
                    com.wuyiming.singlecolumnadfeed_android.data.model.AdInsight insight =
                            gen.generate(item.getTitle(), item.getDescription());
                    item.setInsight(insight);
                } catch (Exception e) {
                    // local fallback handled inside generator
                }
                postFeedState(true);
            }
        });
    }

    public LiveData<FeedUiState> getUiState() { return uiState; }
    public LiveData<String> getLikeSyncEvent() { return likeSyncEvent; }

    @Override
    protected void onCleared() {
        super.onCleared();
        dataExecutor.shutdown();
        searchExecutor.shutdown();
        refreshExecutor.shutdown();
    }
}

File: app/src/main/java/com/wuyiming/singlecolumnadfeed_android/viewmodel/FeedViewModel.kt
package com.wuyiming.singlecolumnadfeed_android.viewmodel

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.wuyiming.singlecolumnadfeed_android.ai.AiInsightGenerator
import com.wuyiming.singlecolumnadfeed_android.ai.AiSearchService
import com.wuyiming.singlecolumnadfeed_android.data.model.FeedCategory
import com.wuyiming.singlecolumnadfeed_android.data.model.FeedItem
import com.wuyiming.singlecolumnadfeed_android.data.repository.DefaultFeedRepository
import com.wuyiming.singlecolumnadfeed_android.data.repository.FeedRepository
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.launch

class FeedViewModel(
    application: Application,
    private val feedRepository: DefaultFeedRepository = DefaultFeedRepository(application)
) : AndroidViewModel(application) {

    companion object {
        private const val PAGE_SIZE = 20
        private const val MIN_REFRESH_ANIMATION_MS = 600L

        @Volatile
        private var sharedInstance: FeedViewModel? = null

        @JvmStatic fun getSharedInstance(app: Application): FeedViewModel {
            return sharedInstance ?: synchronized(FeedViewModel::class.java) {
                sharedInstance ?: FeedViewModel(app).also { sharedInstance = it }
            }
        }

        @JvmStatic fun getInstance(app: Application): FeedViewModel = getSharedInstance(app)
    }

    private val allItems = MutableStateFlow<List<FeedItem>>(emptyList())
    private val selectedCategory = MutableStateFlow(FeedCategory.RECOMMEND)
    private val isRefreshing = MutableStateFlow(false)
    private val isLoadingMore = MutableStateFlow(false)
    private val hasMoreItems = MutableStateFlow(true)
    private val loadMoreState = MutableStateFlow<LoadMoreState>(LoadMoreState.Idle)
    private val currentPage = MutableStateFlow(1)
    private val screenState = MutableStateFlow<FeedScreenState>(FeedScreenState.Loading)
    private val isSearchMode = MutableStateFlow(false)
    private val isSearching = MutableStateFlow(false)
    private var refreshSeed = 0
    private var hasLoadedOnce = false
    private var aiInsightJob: Job? = null

    private val _feedItems = MutableStateFlow<List<FeedItem>>(emptyList())
    val allFeedItems: StateFlow<List<FeedItem>> = allItems.asStateFlow()
    val feedItems: StateFlow<List<FeedItem>> = _feedItems.asStateFlow()
    val currentCategory: StateFlow<FeedCategory> = selectedCategory.asStateFlow()
    val refreshing: StateFlow<Boolean> = isRefreshing.asStateFlow()
    val loadingMore: StateFlow<Boolean> = isLoadingMore.asStateFlow()
    val hasMore: StateFlow<Boolean> = hasMoreItems.asStateFlow()
    val currentLoadMoreState: StateFlow<LoadMoreState> = loadMoreState.asStateFlow()
    val currentScreenState: StateFlow<FeedScreenState> = screenState.asStateFlow()
    val searchMode: StateFlow<Boolean> = isSearchMode.asStateFlow()
    val searching: StateFlow<Boolean> = isSearching.asStateFlow()

    // AI services (nullable = AI disabled when not configured)
    private var aiSearchService: AiSearchService? = null

    init {
        observeFilteredItems()
        refresh()
    }

    fun configureAi(apiKey: String, apiUrl: String, model: String) {
        if (apiKey.isBlank()) return
        feedRepository.aiInsightGenerator = AiInsightGenerator(apiKey, apiUrl, model)
        aiSearchService = AiSearchService(apiKey, apiUrl, model)
        // 对已有数据补生成 AI 摘要
        val snapshot = allItems.value
        if (snapshot.isNotEmpty()) generateInsightsFor(snapshot)
    }

    private fun observeFilteredItems() {
        viewModelScope.launch {
            combine(allItems, selectedCategory) { items, category ->
                items.filter { it.category == category }
            }.collect { filteredItems ->
                _feedItems.value = filteredItems
                if (hasLoadedOnce && screenState.value !is FeedScreenState.Error && screenState.value !is FeedScreenState.Loading) {
                    screenState.value = filteredItems.toScreenState()
                }
            }
        }
    }

    fun refresh() {
        viewModelScope.launch {
            val refreshStart = System.currentTimeMillis()
            val showFullLoading = !hasLoadedOnce || screenState.value is FeedScreenState.Error
            if (showFullLoading) screenState.value = FeedScreenState.Loading
            isRefreshing.value = true
            try {
                aiInsightJob?.cancel()
                refreshSeed += 1
                currentPage.value = 1
                hasMoreItems.value = true
                loadMoreState.value = LoadMoreState.Idle
                val loadedItems = feedRepository.loadFeedItems(1, PAGE_SIZE, refreshSeed)
                allItems.value = loadedItems
                hasLoadedOnce = true
                hasMoreItems.value = loadedItems.size == PAGE_SIZE
                screenState.value = loadedItems.toScreenState()
                generateInsightsFor(loadedItems)
            } catch (e: Exception) {
                if (!hasLoadedOnce) {
                    screenState.value = FeedScreenState.Error(e.message ?: "加载失败")
                }
            } finally {
                val elapsed = System.currentTimeMillis() - refreshStart
                val remaining = MIN_REFRESH_ANIMATION_MS - elapsed
                if (remaining > 0) delay(remaining)
                isRefreshing.value = false
            }
        }
    }

    fun loadMore() {
        if (isLoadingMore.value || !hasMoreItems.value || isSearchMode.value) return
        viewModelScope.launch {
            isLoadingMore.value = true
            loadMoreState.value = LoadMoreState.Loading
            try {
                val nextPage = currentPage.value + 1
                val nextItems = feedRepository.loadFeedItems(nextPage, PAGE_SIZE, refreshSeed)
                allItems.value = allItems.value + nextItems
                currentPage.value = nextPage
                hasMoreItems.value = nextItems.isNotEmpty() && nextPage < 5
                loadMoreState.value = if (hasMoreItems.value) LoadMoreState.Idle else LoadMoreState.EndReached
                generateInsightsFor(nextItems)
            } catch (e: Exception) {
                loadMoreState.value = LoadMoreState.Error(e.message ?: "下一页加载失败")
            } finally {
                isLoadingMore.value = false
            }
        }
    }

    fun switchCategory(category: FeedCategory) {
        selectedCategory.value = category
    }

    fun toggleLike(id: String) {
        allItems.value = allItems.value.map { item ->
            if (item.id == id) feedRepository.toggleLike(item) else item
        }
    }

    fun toggleCollect(id: String) {
        allItems.value = allItems.value.map { item ->
            if (item.id == id) feedRepository.toggleCollect(item) else item
        }
    }

    fun search(query: String) {
        if (query.isBlank()) return
        viewModelScope.launch {
            isSearching.value = true
            isSearchMode.value = true
            val trimmed = query.trim()
            // AI 云端拆词优先，失败降级本地分词
            val keywords = try {
                val service = aiSearchService
                if (service != null && trimmed.length > 2) {
                    service.extractKeywords(trimmed)
                } else {
                    splitLocal(trimmed)
                }
            } catch (_: Exception) {
                splitLocal(trimmed)
            }
            val results = allItems.value.filter { item ->
                keywords.any { kw ->
                    item.title.contains(kw, ignoreCase = true) ||
                    item.description.contains(kw, ignoreCase = true) ||
                    item.insight?.tags?.any { it.contains(kw, ignoreCase = true) } == true
                }
            }
            _feedItems.value = results
            isSearching.value = false
            screenState.value = results.toScreenState()
        }
    }

    fun clearSearch() {
        isSearchMode.value = false
        isSearching.value = false
        _feedItems.value = allItems.value.filter { it.category == selectedCategory.value }
        screenState.value = _feedItems.value.toScreenState()
    }

    // ---- AI insight generation (background, non-blocking) ----

    private fun generateInsightsFor(items: List<FeedItem>) {
        val gen = feedRepository.aiInsightGenerator ?: return
        aiInsightJob = viewModelScope.launch {
            for (item in items) {
                // 跳过已有 summary 的项
                if (item.insight?.summary != null && item.insight.summary.isNotEmpty()) continue
                try {
                    val insight = gen.generate(item.title, item.description)
                    if (insight != null) {
                        allItems.value = allItems.value.map { current ->
                            if (current.id == item.id) current.toBuilder().insight(insight).build() else current
                        }
                    }
                } catch (_: Exception) {
                    // local fallback handled inside generator
                }
            }
        }
    }

    private fun splitLocal(query: String): List<String> {
        val parts = query.split("[\\s,.!?;:，。！？；：]+".toRegex()).filter { it.isNotEmpty() }
        return parts.ifEmpty { listOf(query) }
    }

    private fun List<FeedItem>.toScreenState(): FeedScreenState {
        return if (isEmpty()) FeedScreenState.Empty else FeedScreenState.Content
    }
}

sealed interface FeedScreenState {
    data object Loading : FeedScreenState
    data object Content : FeedScreenState
    data object Empty : FeedScreenState
    data class Error(val message: String) : FeedScreenState
}

sealed interface LoadMoreState {
    data object Idle : LoadMoreState
    data object Loading : LoadMoreState
    data object EndReached : LoadMoreState
    data class Error(val message: String) : LoadMoreState
}
